feat(install): warn if default linker (`cc`) is missing from PATH

cachebag opened the pull request:
Closes #4356 

Test was run in the same environment outlined in the original issue (using `docker run --rm -it debian:testing bash`) and simulates an environment where `cc` is not installed.
<br>
<img width="730" height="563" alt="image" src="https://github.com/user-attachments/assets/1978fe0a-fc6a-43f4-abd9-ae9612a3d250" />

cachebag committed:
feat(install): warn if default linker (cc) is missing; add respective test case

djc commented on src/cli/self_update.rs:
```diff
@@ -742,6 +744,20 @@ fn current_install_opts(opts: &InstallOpts<'_>, process: &Process) -> String {
+            warn!("No default linker (`cc`) was found in your PATH.");
+            warn!("Many Rust crates require a system C toolchain to build.");
+            warn!("For example, on Debian/Ubuntu you can install it with:");
+            warn!("    sudo apt install build-essential");
```
I don't think we want to give distro-specific advise in our log messages.

cachebag replied:
Makes sense, thanks. Fixed

djc reviewed (approved):
LGTM modulo a small nit, thanks!

rami3l commented on src/cli/self_update.rs:
```diff
@@ -742,6 +744,18 @@ fn current_install_opts(opts: &InstallOpts<'_>, process: &Process) -> String {
     )
 }
 
+fn warn_if_default_linker_missing(_process: &Process) {
```
It looks like the `_process` param is unused, what about removing it?

cachebag replied:
I ended up using it via your other suggestion- initially I felt it right to keep the code consistent but removing it makes sense

rami3l commented on src/cli/self_update.rs:
```diff
@@ -742,6 +744,18 @@ fn current_install_opts(opts: &InstallOpts<'_>, process: &Process) -> String {
 
+fn warn_if_default_linker_missing(_process: &Process) {
+    match Command::new("cc").arg("--version").output() {
+        Ok(output) if output.status.success() => {
```
I thought the [original suggestion](https://github.com/rust-lang/rustup/issues/4356#issuecomment-2928041615) said checking the `PATH` for `cc`, not running `cc`. I thus suggest extracting the relevant function from [here](https://github.com/rust-lang/rustup/blob/fc0a76bd280293da1bdb8fab484d6468109d6ccd/src/toolchain.rs#L416-L434).

rami3l replied:
Also, this makes your new test case actually workable because you can override `PATH` in the test...

cachebag replied:
I see what you mean, okay I redid the method. Thank you for pointing this out

rami3l replied:
@cachebag Hmmm sorry for not being clear enough, but the point here is not to use `Command::new()`; rather in the link [here](https://github.com/rust-lang/rustup/blob/fc0a76bd280293da1bdb8fab484d6468109d6ccd/src/toolchain.rs#L416-L434) there is a piece of code that searches a bin directly inside `PATH` and you can kind of copy and modify to look for `cc` here.

rami3l replied:
After a second look I concluded that extracting a function from there was not necessary, but you can still use `env::split_paths(&path)` with `path` fetched from `process.env_os("PATH")`.

rami3l commented on src/cli/self_update.rs:
```diff
@@ -742,6 +744,18 @@ fn current_install_opts(opts: &InstallOpts<'_>, process: &Process) -> String {
+            // Found a working `cc`, nothing to do.
+        }
+        _ => {
+            warn!("No default linker (`cc`) was found in your PATH.");
```
For consistency, don't put capitals at the beginning or dots at the end.

ChrisDenton commented:
This shouldn't run on Windows. We already have code to handle a missing linker (which is called `link` not `cc`).

Also I'm assuming `cc` is the default for all other hosts but I don't know for certain if that's true.

rami3l commented:
> This shouldn't run on Windows. We already have code to handle a missing linker (which is called `link` not `cc`).
> 
> Also I'm assuming `cc` is the default for all other hosts but I don't know for certain if that's true.

@ChrisDenton I have second thoughts as well, since (if I am reading it right) ~~starting from Rust v1.90 cc will no longer be a hard requirement on Linux x64 (edit: GNU ABI) because it will use the builtin linker instead, making this PR obsolete~~:

~~https://blog.rust-lang.org/2025/09/01/rust-lld-on-1.90.0-stable/~~

~~Considering most users who have encountered this issue are indeed using Linux x64, I'm afraid the best option in this case is to do nothing.~~

Edit: The reasoning is inexact. See https://github.com/rust-lang/rustup/pull/4469#issuecomment-3251983199.

cachebag commented:
@rami3l 

What about non-x86_64 Linux systems? It's probably like 1% of Linux users, so if that's not pertinent due to `lld` fixing the other 99% of cases I'll close this, it makes sense. 

I didn't know that, that's interesting.

rami3l commented:
@Kobzol I just tried the latest beta release and it looks like `cc` is still required; am I misunderstanding something here regarding the new default linker thing?

Also cc https://github.com/rust-lang/rust/issues/46998.

Kobzol commented:
Linking is still done through `cc` on Linux, that's just kind of how it works :sweat_smile: It's just that the actually used linker is LLD, and not BFD.

lqd commented:
`cc` is still the *linker driver*, yup, it knows how to pass the additional object files from the system. It's not like windows where link.exe or lld.exe are the `-Clinker`, on linux it's just some additional configuration to `-Clinker=cc` to tell *it* which linker to use.

rami3l commented:
@Kobzol @lqd Thanks for the clarification :)

@cachebag I'm really sorry for the false alarm, would you mind pushing this patch towards the finish line? 🙏

cachebag commented:
> @Kobzol @lqd Thanks for the clarification :)
> 
> 
> 
> @cachebag I'm really sorry for the false alarm, would you mind pushing this patch towards the finish line? 🙏 

@rami3l 
Yes no worries. I did make a few changes based on your suggestions- removed the windows check and just extracted the function you mentioned, which are posted. Was there anything more that should be done beyond that?

rami3l commented:
> Yes no worries. I did make a few changes based on your suggestions- removed the windows check and just extracted the function you mentioned, which are posted. Was there anything more that should be done beyond that?

@cachebag Did you forget to push? Because I see the current version in an incoherent state, e.g. no special casing for Windows, no updated test snapshot, ...

cachebag commented:
> > Yes no worries. I did make a few changes based on your suggestions- removed the windows check and just extracted the function you mentioned, which are posted. Was there anything more that should be done beyond that?
> 
> @cachebag Did you forget to push? Because I see the current version in an incoherent state, e.g. no special casing for Windows, no updated test snapshot, ...

Should be good now- my mistake I did forget to push earlier but I adjusted it now to match the behavior of the test case I wrote for this patch.

rami3l commented on src/cli/self_update.rs:
```diff
@@ -610,6 +610,8 @@ pub(crate) async fn install(
     };
     md(&mut term, msg);
 
+    warn_if_default_linker_missing(process);
```
As @ChrisDenton has mentioned, there is no need to run this code in Windows, so `cfg(unix)` for both the declaration and the call.

rami3l commented on src/cli/self_update.rs:
```diff
@@ -742,6 +744,30 @@ fn current_install_opts(opts: &InstallOpts<'_>, process: &Process) -> String {
+            p.push(&cc_binary);
+            if p.is_file() {
+                // Try to run the found `cc` to see if it works
+                match Command::new(&p).arg("--version").output() {
```
I believe there is no need to pen test `cc` here with `Command::new()`; the actual Rust project will do the real thing for us and we only need to emit a warning indicating a suspicious behavior.

rami3l commented on tests/suite/cli_inst_interactive.rs:
```diff
@@ -626,3 +626,30 @@ warn: It looks like you have an existing rustup settings file at:
+...
+warn: No default linker (`cc`) was found in your PATH.
+warn: Many Rust crates require a system C toolchain to build.
+warn: For example, on Debian/Ubuntu you can install it with:
```
The snapshot is still out of date.

rami3l commented on tests/suite/cli_inst_interactive.rs:
```diff
@@ -626,3 +626,30 @@ warn: It looks like you have an existing rustup settings file at:
 "#]]);
 }
+
+#[tokio::test]
```
Also, the changes are simple enough so a single commit will suffice.

rami3l reviewed (approved):
@cachebag Many thanks :)

rami3l commented:
> `cc` is still the _linker driver_, yup, it knows how to pass the additional object files from the system. It's not like windows where link.exe or lld.exe are the `-Clinker`, on linux it's just some additional configuration to `-Clinker=cc` to tell _it_ which linker to use.

@lqd Just to be sure, do you happen to know where can I find this information about the platform-linker driver correspondence? I'm still a bit afraid of emitting the warning for more platforms than we should 🤔

rami3l commented on tests/suite/cli_inst_interactive.rs:
```diff
@@ -626,3 +626,28 @@ warn: It looks like you have an existing rustup settings file at:
 "#]]);
 }
+
+#[tokio::test]
```
This test also needs to be unix-only.

lqd commented:
> do you happen to know where can I find this information about the platform-linker driver correspondence

Hum. In general, the target `compiler/rustc_target/src/spec` themselves have information that should give the default linker, but that can just be the linker flavor in particular. That can of course be overridden on the CLI...

I don't know if we have such a thing, but https://doc.rust-lang.org/nightly/nightly-rustc/rustc_target/spec/enum.LinkerFlavor.html should give you an idea.

rami3l commented:
@lqd Thanks for the explanation!

From the link that you have mentioned above I conclude that the false alarms, if they do exist, are pretty much constrained given that we don't interface with WASM, Emscripten, Nvidia PTX and similar, so it should in most cases be a simple Windows/Unix distinction.

As such, I am merging this PR first to see how it goes.
